Mathlib/RingTheory/Jacobson/Artinian.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Artinian.Ring
+import Mathlib.RingTheory.HopkinsLevitzki
+import Mathlib.RingTheory.Jacobson.Ring
+
+/-!
+# Artinian rings over jacobson rings
+
+## Main results
+- `Module.finite_iff_isArtinianRing`: If `A` is a finite type algebra over an artinian ring `R`,
+then `A` is finite over `R` if and only if `A` is an artinian ring.
+
+-/
+
+variable {R A : Type*} [CommRing R] [CommRing A] [Algebra R A] [Algebra.FiniteType R A]
+
+lemma Module.finite_of_finite_mul_of_field [IsJacobsonRing R]
+    (I J : Ideal A) [I.IsMaximal]
+    [IsArtinian A J] [Module.Finite R (I * J)] : Module.Finite R J := by
+  let IJ := Submodule.comap J.subtype (I * J)
+  have : Module.IsTorsionBySet A (J ⧸ IJ) I := by
+    rintro x ⟨y, hy : y ∈ I⟩
+    obtain ⟨⟨x, hx⟩, rfl⟩ := Submodule.mkQ_surjective _ x
+    rw [← map_smul]
+    simpa [Submodule.Quotient.mk_eq_zero] using Ideal.mul_mem_mul hy hx
+  letI : Module (A ⧸ I) (J ⧸ IJ) := this.module
+  letI := Ideal.Quotient.field I
+  haveI : Module.Finite R (A ⧸ I) := finite_of_finite_type_of_isJacobsonRing R (A ⧸ I)
+  have : Module.Finite R (J ⧸ IJ) :=
+    have := ((IsArtinianRing.tfae (A ⧸ I) (J ⧸ IJ)).out 2 0 rfl rfl).mp
+      (isArtinian_of_tower A inferInstance)
+    .trans (A ⧸ I) _
+  have : Module.Finite R J := by
+    constructor
+    apply Submodule.fg_of_fg_map_of_fg_inf_ker (IJ.mkQ.restrictScalars R)
+    · rwa [Submodule.map_top, LinearMap.range_eq_top.mpr, ← Module.finite_def]
+      exact IJ.mkQ_surjective
+    · simp only [LinearMap.ker_restrictScalars, Submodule.ker_mkQ, le_top, inf_of_le_right]
+      apply Submodule.fg_of_fg_map_injective (J.subtype.restrictScalars R) Subtype.val_injective
+      rw [← Submodule.restrictScalars_map, Submodule.map_comap_eq_of_le, ← Submodule.fg_top]
+      · exact Module.Finite.fg_top (M := I * J)
+      · rw [Submodule.range_subtype]; exact Ideal.mul_le_left
+  exact this
+
+/-- If `A` is a finite type algebra over `R`, then `A` is an artinian ring and `R` is jacobson
+implies `A` is finite over `R`. -/
+lemma Module.finite_of_isArtinianRing [IsJacobsonRing R] [IsArtinianRing A] :
+    Module.Finite R A := by
```

Comment:
I think you can refer to https://stacks.math.columbia.edu/tag/0CY7 since this result is a strict generalization (Field -> ArtinianRing). Do you have a reference for this exact result? Wikipedia calls 0CY7 [Zariski's lemma](https://en.wikipedia.org/wiki/Zariski%27s_lemma). I got curious whether Field can be replaced by DivisionRing and found that it's an open problem: see https://mathoverflow.net/questions/324885/infinite-dimensional-finitely-generated-algebraic-division-algebra
![image](https://github.com/user-attachments/assets/0c1ab567-b04e-4156-8cc8-e5a34a79ba32)

Reply:
I think that tag goes to `finite_of_finite_type_of_isJacobsonRing`, which is a key ingredient of the proof.

I can't find the proof I am following (it was done 3 years ago), It might be a paraphrase of Atiyah Mccdonald Exercise 8.3 (which says that artinian <=> k-finite for an algebra of finite type over a field k), but the proof doesn't really match the hint. The proof of 6.11 might be the inspiration instead.

This lemma is needed in 1 => 2 of [stacks#00PJ](https://stacks.math.columbia.edu/tag/00PJ) and in there they pointed directly to Noether normalization for the proof of "artinian => k-finite".

Reply:
Yeah, the Wikipedia page on Zariski's lemma also says that one proof uses Noether normalization (together with that integral extensions preserve Krull dimension). I don't think the existing proof in mathlib uses it, since Noether normalization hasn't been merged yet ...

Reply:
@alreadydone Is there anything that you suggest I add?

Reply:
I suggest adding a reference to Zariski's lemma and stacks#0CY7 at [finite_of_finite_type_of_isJacobsonRing](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Jacobson/Ring.html#finite_of_finite_type_of_isJacobsonRing), not necessarily in this PR.

In [this commit](https://github.com/leanprover-community/mathlib4/commit/69a33893e5af81cf631a10154b652db5b6c1697f#diff-6adf2b9b13e4ed13e1e69a2f2e0de9f64952a597413d648204ed9961d26b9e6eL35-R38) I show how to refactor the proof of `IsSemiprimaryRing.isNoetherian_iff_isArtinian` to accommodate the result in this PR. Maybe you and @jcommelin can judge which approach is better.

Reply:
It seems like your approach gives some stronger result as well? I don't think I would need those results but It's nice to have them anyways.

Reply:
Perhaps It would be faster if you just open a PR on [your branch](https://github.com/leanprover-community/mathlib4/compare/artinianJacobson_demo)? But it would also work if you just merge that branch into this PR.

Reply:
Okay, opened #21904